scripts/autolabel.lean

```diff
@@ -0,0 +1,195 @@
+
+/--
+Return all labels names for labels in `mathlibLabels` which are matching
+at least one of the `files`.
+
+* `files`: array of relative paths starting from the mathlib project directory.
+-/
+def getMatchingLabels (files : Array FilePath) : Array String :=
+  let applicable := mathlibLabels.filter fun label ↦
+    -- first exclude all files the label excludes,
+    -- then see if any file remains included by the label
+    let notExcludedFiles := files.filter fun file ↦
+      label.exclusions.all (!·.isPrefixOf file)
+    label.dirs.any (fun dir ↦ notExcludedFiles.any (dir.isPrefixOf ·))
+  -- return sorted list of label names
+  applicable.map (·.label) |>.qsort (· < ·)
+
+/-!
+Testing the functionality of the declarations defined in this script
+-/
+section Tests
+
+-- Test `FilePath.isPrefixOf`
+#guard ("Mathlib" / "Algebra" : FilePath).isPrefixOf ("Mathlib" / "Algebra" / "Basic.lean")
+
+-- Test `FilePath.isPrefixOf` does not trigger on partial prefixes
+#guard ! ("Mathlib" / "Algebra" : FilePath).isPrefixOf ("Mathlib" / "AlgebraicGeometry")
+
+#guard getMatchingLabels #[] == #[]
+-- Test default value for `label.dirs` works
+#guard getMatchingLabels #["Mathlib" / "SetTheory" / "ZFC"] == #["t-set-theory"]
+-- Test exclusion
+#guard getMatchingLabels #["Mathlib" / "Tactic"/ "Abel.lean"] == #["t-meta"]
+#guard getMatchingLabels #["Mathlib" / "Tactic"/ "Linter" / "Lint.lean"] == #["t-linter"]
+#guard getMatchingLabels #[
+  "Mathlib" / "Tactic"/ "Linter" / "Lint.lean",
+  "Mathlib" / "Tactic" / "Abel.lean" ] == #["t-linter", "t-meta"]
+
+end Tests
+
+end AutoLabel
+
+open IO AutoLabel in
+
+/-- `args` is expected to have length 1, and the first argument is the PR number. -/
+unsafe def main (args : List String): IO Unit := do
+  if args.length > 1 then
+    println s!"autolabel: invalid number of arguments ({args.length}). Please run without \
+    arguments or provide the target PR's number as single argument!"
+    IO.Process.exit 1
+  let prNumber? := args[0]?
+
+  -- validate that all paths in `mathilbLabels` actually exist
+  let mut valid := true
+  for label in mathlibLabels do
+    for dir in label.dirs do
+      unless ← FilePath.pathExists dir do
+        println s!"error: directory {dir} does not exist! (from label {label.label})"
+        valid := false
+    for dir in label.exclusions do
+      unless ← FilePath.pathExists dir do
+        println s!"error: excluded directory {dir} does not exist! (from label {label.label})"
+        valid := false
+  unless valid do
+    IO.Process.exit 2
+
+  -- get the modified files
+  let gitDiff ← IO.Process.run {
+    cmd := "git",
+    args := #["diff", "--name-only", "origin/master...HEAD"] }
+  let modifiedFiles : Array FilePath := (gitDiff.splitOn "\n").toArray.map (⟨·⟩)
+
+  -- find labels covering the modified files
+  let labels := getMatchingLabels modifiedFiles
+
+  match labels with
+  | #[] =>
+    println s!"No applicable labels found!"
```

Comment:
I like the validation check you added. I meant to (also) throw an error if this branch is reached --- unless the PR modifies e.g. only `scripts`.

Reply:
I'm a bit hesitant to make this failing for the following reasons:

- it will mean the PR gets a red cross because there was no applicable label, even if the PR is passing CI. (afaik there is no way to make CI pass despite a single workflow failing).
- There are still plenty of folders/files which are not covered by any label. For example most things in `Counterexamples`, `Archive`, in the root dir, `Mathilb/[Testing,Deprecated,Init]`. It seems unreasonable to me that the CI should fail on any PR modifying these labels.

However, I did add a test ensuring that all subfolders of `Mathlib/` are covered by a label, which might address your concerns in a way?

Reply:
Good points - makes sense to not make CI fail. The current test is good enough for me!